### ISSUE TITLE
haskellPackages: Build failure on ghc927 but not ghc928

### ISSUE THREAD

**Issue by shlok** [NixOS/nixpkgs#240512]
### Describe the bug

My Haskell project (see below) compiles fine when using `final.haskell.packages.ghc928`, but when I switch to `final.haskell.packages.ghc927`, it fails with the error 

```txt
       > Configuring streamly-lmdb-0.7.0...
       >
       > Setup: Encountered missing or private dependencies:
       > streamly >=0.9 && <0.10

```

#### 1
Please create a new directory named `my-test` and put the following files in there:

`my-test.cabal`:
```cabal
cabal-version:      3.0
name:               my-test
version:            0.1.0.0
build-type:         Simple

executable my-test
    main-is: Main.hs
    default-language: Haskell2010
    hs-source-dirs: app
    build-depends:
      , base >=4.7 && <5
      , streamly >=0.9 && <0.10
      , streamly-core >=0.1 && <0.2
      , streamly-lmdb ==0.7.0
```

`app/Main.hs`:
```haskell
module Main where

main :: IO ()
main = putStrLn "hello"
```

`flake.lock`
```txt
{
  "nodes": {
    "flake-utils": {
      "inputs": {
        "systems": "systems"
      },
      "locked": {
        "lastModified": 1687709756,
        "narHash": "sha256-Y5wKlQSkgEK2weWdOu4J3riRd+kV/VCgHsqLNTTWQ/0=",
        "owner": "numtide",
        "repo": "flake-utils",
        "rev": "dbabf0ca0c0c4bce6ea5eaf65af5cb694d2082c7",
        "type": "github"
      },
      "original": {
        "owner": "numtide",
        "repo": "flake-utils",
        "type": "github"
      }
    },
    "nixpkgs": {
      "locked": {
        "lastModified": 1687886075,
        "narHash": "sha256-PeayJDDDy+uw1Ats4moZnRdL1OFuZm1Tj+KiHlD67+o=",
        "owner": "NixOS",
        "repo": "nixpkgs",
        "rev": "a565059a348422af5af9026b5174dc5c0dcefdae",
        "type": "github"
      },
      "original": {
        "owner": "NixOS",
        "ref": "nixpkgs-unstable",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "root": {
      "inputs": {
        "flake-utils": "flake-utils",
        "nixpkgs": "nixpkgs"
      }
    },
    "systems": {
      "locked": {
        "lastModified": 1681028828,
        "narHash": "sha256-Vy1rq5AaRuLzOxct8nz4T6wlgyUR7zLU309k9mBC768=",
        "owner": "nix-systems",
        "repo": "default",
        "rev": "da67096a3b9bf56a91d16901293e51ba5b49a27e",
        "type": "github"
      },
      "original": {
        "owner": "nix-systems",
        "repo": "default",
        "type": "github"
      }
    }
  },
  "root": "root",
  "version": 7
}
```

`flake.nix`
```nix
{
  description = "my-test";

  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
    flake-utils.url = "github:numtide/flake-utils";
  };

  outputs = { self, nixpkgs, flake-utils }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        ghcVersion = "928";
        config = {};

        overlays = [
          (final: prev:
            let
              haskellPkgs = final.haskell.packages."ghc${ghcVersion}";
            in {
              myHaskellPkgs = haskellPkgs.override {
                overrides = hfinal: hprev: {
                  my-test =
                    hfinal.callCabal2nix "my-test" ./. {
                      streamly = hfinal.streamly_0_9_0;
                    };
                };
              };

              my-test = final.myHaskellPkgs.my-test;

              myDevShell = final.myHaskellPkgs.shellFor {
                packages = p: [
                    p.my-test
                  ];

                buildInputs = [];

                nativeBuildInputs = [
                  final.myHaskellPkgs.cabal-install
                  final.myHaskellPkgs.haskell-language-server
                ];
              };
            })
        ];

        pkgs = import nixpkgs { inherit config overlays system; };
      in {
        packages = {
          default = pkgs.my-test;
          my-test = pkgs.my-test;
        };
        
        devShells.default = pkgs.myDevShell;
      });
}
```

#### 2
`nix develop -L` should work fine.

#### 3
In `flake.nix`, please change `"928"` to `"927"` and run `nix develop -L`  again. It should now fail with the aforementioned error.

### Additional information

* The `flake.lock` above refers to the latest `nixpkgs-unstable` (at the time of this writing), where `hackage-packages.nix` still has `"streamly"` (0.8.x) and `"streamly_0_9_0"` as separate packages. (This is in contrast to the latest `haskell-updates` branch, where `"streamly"` is now 0.9.0 and `"streamly_0_9_0"` has been removed.)
* Note that `configuration-common.nix` and `configuration-nix.nix` have `streamly-lmdb`-specific fixes. Are they for some reason working fine for `"928"` but not for to `"927"`?

CC @maralorn @sternenseemann

**Comment by sternenseemann** [comment]
Turns out the bug was of your own making: a1a4c4d6296ec93e72dac23df6965681e384af74. Fix at #240616.

For the future, please try reducing the reproducer first—for reproducing the issue it is enough to build `haskell.packages.ghc928.streamly` for the nixpkgs commit given for you. Your reproducer includes `callCabal2nix`, flake-utils, `shellFor`, a custom overlay and all sorts of things that need to be untangled first before the issue is pinned down. The log already gives a solid hint where that is.

**Comment by shlok** [comment]
@sternenseemann Nice catch, and thanks for the feedback!
@maralorn No worries. I should have seen this, and in the future I will test things on a non-default Haskell set too.
